pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,404 @@
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewL7VisibilityController(
+	ofClient openflow.Client,
+	interfaceStore interfacestore.InterfaceStore,
+	podInformer cache.SharedIndexInformer,
+	namespaceInformer coreinformers.NamespaceInformer,
+	l7Reconciler *l7engine.Reconciler) *L7VisibilityController {
+	podToDirectionMap := make(map[string]v1alpha2.Direction)
+	l7c := &L7VisibilityController{
+		ofClient:              ofClient,
+		interfaceStore:        interfaceStore,
+		podInformer:           podInformer,
+		podLister:             corelisters.NewPodLister(podInformer.GetIndexer()),
+		podListerSynced:       podInformer.HasSynced,
+		namespaceInformer:     namespaceInformer.Informer(),
+		namespaceLister:       namespaceInformer.Lister(),
+		namespaceListerSynced: namespaceInformer.Informer().HasSynced,
+		l7Reconciler:          l7Reconciler,
+		PodToDirectionMap:     podToDirectionMap,
+		L7SuricataSocketPath:  socketPath,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "l7VisibilityController"),
+	}
+	l7c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    l7c.addPod,
+			UpdateFunc: l7c.updatePod,
+			DeleteFunc: l7c.deletePod,
+		},
+		resyncPeriod,
+	)
+	l7c.namespaceInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    l7c.addNamespace,
+			UpdateFunc: l7c.updateNamespace,
+		},
+		resyncPeriod,
+	)
+	return l7c
+}
+
+func (l7c *L7VisibilityController) Run(stopCh <-chan struct{}) {
+	defer l7c.queue.ShutDown()
+	klog.InfoS("Starting", "Controller", controllerName)
+	defer klog.InfoS("Shutting down", "Controller", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, l7c.podListerSynced, l7c.namespaceListerSynced) {
+		return
+	}
+	if itf, ok := l7c.interfaceStore.GetInterfaceByName(config.L7NetworkPolicyTargetPortName); ok {
+		l7c.targetPort = uint32(itf.OFPort)
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(l7c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (l7c *L7VisibilityController) worker() {
+	for l7c.processNextWorkItem() {
+	}
+}
+
+func (l7c *L7VisibilityController) processNextWorkItem() bool {
+	obj, quit := l7c.queue.Get()
+	if quit {
+		return false
+	}
+	defer l7c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		// As the item in the work queue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		l7c.queue.Forget(key)
+		klog.ErrorS(nil, "Expected string in work queue but got", "key", obj)
+		return true
+	} else if err := l7c.syncPod(key); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until
+		// another change happens.
+		l7c.queue.Forget(key)
+	} else if err == errInvalidAnnotation || err == errPodInterfaceNotFound {
```

Comment:
I'm confused as to why `errPodInterfaceNotFound` is handled this way (item forgotten), maybe you could add a comment in the code describing the reason?

Reply:
added!

Reply:
The comment describes what the code is doing, and therefore is not very useful. I was asking why specifically we ignore `errPodInterfaceNotFound`.

For `errInvalidAnnotation`, it is pretty straightforward to me (no harm in including this information in a comment though): there is no point in requeuing, and after the user fixes the annotation by updating the Pod, our event handler will be called again and we will process the Pod again. 

However, for `errPodInterfaceNotFound`, I don't know what the rationale is?

Reply:
is your concern about why do we not re enqueue it ?

Reply:
I understood the first bullet, however for the second, if the Pod has an IP and it does not have interface, should we not re-enqueue it ? my understanding was if the Pod receives an IP it should also have an interface thus I was not re-enqueueing it.

Reply:
Thanks @tnqn , added code to re-enqueue the key if the pod ip is present but interface is absent